core/src/main/scala/org/apache/spark/deploy/master/WorkerInfo.scala

```diff
@@ -89,17 +122,52 @@ private[spark] class WorkerInfo(
     drivers(driver.id) = driver
     memoryUsed += driver.desc.mem
     coresUsed += driver.desc.cores
+    driverToResourcesUsed(driver.id) = driver.resources
   }
 
   def removeDriver(driver: DriverInfo) {
     drivers -= driver.id
     memoryUsed -= driver.desc.mem
     coresUsed -= driver.desc.cores
+    driverToResourcesUsed.remove(driver.id)
   }
 
   def setState(state: WorkerState.Value): Unit = {
     this.state = state
   }
 
   def isAlive(): Boolean = this.state == WorkerState.ALIVE
+
+  /**
+   * acquire specified amount resources for driver/executor from the worker
+   * @param resourceReqs the resources requirement from driver/executor
+   * @return
+   */
+  def acquireResources(resourceReqs: Map[String, Int])
+    : Map[String, Seq[String]] = {
+    resourceReqs.map { case (rName, amount) =>
+      // TODO (wuyi) rName does not exists ?
+      rName -> resources(rName).acquire(amount)
+    }
+  }
+
+  /**
+   * used during master recovery
+   */
+  def notifyResources(expected: Map[String, Seq[String]]): Unit = {
+    expected.foreach { case (rName, addresses) =>
+      resources(rName).acquire(addresses)
```

Comment:
maybe I"m missing it here, if the master is recovered, aren't the the workers still running possibly?  If you then ask them to acquire but it already has it, it will throw an exception.

Reply:
In HA, we persists `WorkInfo` into `persistenceEngine` once a Worker registered.  And at that time, the `WorkerInfo` has no assigned resources but all free resources. When master recovered, it reads the `WorkInfo` from `persistenceEngine`, so the `WorkInfo` should still be available for all resources.